### ISSUE TITLE
[DoctrineMessenger, Postgres] Postgres LISTEN/NOTIFY improvement proposition for better handling of time limit, multi-queue worker and delayed tasks

### ISSUE THREAD

**Issue by d-ph** [symfony/symfony#46862]
### Description

Hello,

I'm evaluating a use case where queue workers are running with a 10-minute "get_notify_timeout" option (i.e. the max time the [pgsqlGetNotify()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L82) will synchronously wait for notifications from Postgres), and I found a few problems with it, to which I would like to propose a solution. The problems and solutions are interconnected, so it's perhaps best to discuss them in one ticket and then, if needed, create separate tickets/PRs (which I can do when time comes).

# Problem no. 1 (critical)

Imagine there are two Doctrine queues: `high_priority` and `low_priority`, and a queue worker is started with the following cli command:
```
bin/console messenger:consume high_priority low_priority
```
In other words, the worker will consume messages from both queues, preferring the high_priority. Steps to reproduce the problem:

1. Have no tasks queued for both queues.
2. Start the worker.
    1. Due to an [implementation detail](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L62), the following happens:
        1. The worker will conclude that "high_priority" queue is empty and make a note of it. It will not sync-wait on postgres' notifications at this point yet.
        2. The worker will conclude that "low_priority" queue is empty and make a note of it.
        3. The worker enters a new "[master while](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L94)" loop.
        4. While attempting to get a new queue task for the "high_priority" queue, the worker executes a blocking sync-wait for a postgres notification [here](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L82)
    2. It's important to let the worker block at the sync-wait before continuing with these "steps to reproduced". In practice, one should wait for about a second.
3. Add a new task for the "low_priority" queue. One can do it by hand via an INSERT sql statement with appropriate data.
4. Postgres delivers a notification to the worker that a new record has been added to the messenger_queue db table.
5. The worker "wakes up" from the sync-wait.
6. The worker concludes that the new record is not for the "high_priority" queue and continues to the next queue.
7. The worker attempts to retrieve a task for the "low_priority" queue. Due to the fact that [the "queueEmptiedAt" hasn't been cleared for the "low_priority" queue connection](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L62), the worker enters a blocking sync-wait on the "low_priority" queue.

In the last step, the worker should have retrieved the new task for the "low_priority" queue, but instead it entered a blocking sync-wait. This is a problem, especially when the timeout for the sync-wait is 10 minutes, as I mentioned before.

# Problem no. 2

The worker's "--time-limit" option doesn't work as expected, when the worker is in the process of sync-waiting for new queue tasks.

# Problem no. 3

In case there is a delayed task queued that should be picked up for processing after the next 1 minute passes, and the worker enters a sync-wait with a timeout of 10 minutes, then the delayed task will be processed only after the worker times out the sync-wait (assuming nothing else wakes up the worker from the sync-wait).

# Improvement proposition no. 1

I gave these problems a few thoughts and would like to propose the following changes/improvements.

The sync-wait should not happen in [PostgreSqlConnection::get()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L60). Instead, it should happen during the [WorkerRunningEvent](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L128). In other words, conceptually, **the worker should not sync-wait while attempting to retrieve new tasks, but instead should sync-wait while being idle** (i.e. in the [idle sleep](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L127) if-block). In further other words, **instead of the sync-wait attempting to answer the question of "has potentially a new task been added for queue X", it should answer the question of "has something interesting happen on the messenger_queue db table"**.

This would effectively move the majority of the [PostgreSqlConnection::get()](https://github.com/symfony/doctrine-messenger/blob/9695a3eac87d4fe740d74c6f6ff7912a1cdd5588/Transport/PostgreSqlConnection.php#L60) code to an event listener, and would allow the `PostgreSqlConnection::get()` to not override the base class implementation. I believe it would also allow for [the idiosyncrasy in the base class](https://github.com/symfony/doctrine-messenger/blob/8881ca77e18b4f7d52083811462e1f3a34ad107b/Transport/Connection.php#L193) to be removed.

This would fix the problem no. 1 mentioned above. It would also allow for an easier fix for problems no. 2 and no. 3 (see the next improvement proposition).

# Improvement proposition no. 2

With the improvement proposition no. 1 in place, it would allow the worker to choose the sync-wait timeout more intelligently. I propose that the sync-wait timeout value for the `pgsqlGetNotify()` to be chosen in the following way:

1. Query the queue db table to see if there is at least one pending queue task for the queues that the worker has been started for. If yes, do not sync-wait at all.
    1. This potentially prevents the worker from being pwned by the [postgres' notify grouping behaviour](https://github.com/symfony/doctrine-messenger/blob/8881ca77e18b4f7d52083811462e1f3a34ad107b/Transport/Connection.php#L193), but on second thought, it might not be needed at all, considering the fact that the worker's [idle-wait if-block](https://github.com/symfony/messenger/blob/315d93498bc11c22ccf9ba9fa436380488b8fb9a/Worker.php#L127) isn't executed when there was at least 1 task processed in the past while-loop.
2. Query the queue db table to see if there are any delayed queue tasks for the queues that the worker has been started for. Pick the one that becomes available for processing soonest.
3. Compute the worker's remaining time, as requested by the `--time-limit` cli command option.
4. Out of the 3 time values (the soonest delayed task, the remaining --time-limit, and the configured `get_notify_timeout` option) choose the lowest one, and use it for the `pgsqlGetNotify()` sync-wait call.

This fixes problems no. 2 and no. 3. It perhaps is also simpler conceptually, as the sync-waiting logic is cleanly isolated in its own symfony event listener, rather than being a part of an overridden method in `PostgreSqlConnection::get()`.

# Closing words

Implementing the sync-waiting as proposed above would also conceptually usher in a solid opinion on how sync-waiting should be done in Messenger (compared to polling), i.e. the sync-waiting transport should be recommended to sync-wait on any change happening on the underlying queue(s), rather than only sync-waiting on one queue's new tasks.

I truly believe that this is the way to move forward with this, and for it to have an immense value for the ecosystem (one value being that --time-limit, multi-queue workers, and delayed queue tasks would JustWorkTM (properly), even if the worker is in the process of sync-waiting). I can prepare the code change if the proposition is accepted.

Thank you.

### Example

No real example, as the ticket proposes an underlying implementation change, rather than dev-facing interface change (although, the change allows for more features in Messenger to work, when sync-waiting).

**Comment by fabpot** [comment]
@d-ph Thank you for the very detailed issue. The next step would be to see the code changes for the first improvement you suggest here. Would you like to work on a PR?

**Comment by d-ph** [comment]
Hi,

I prepared a code change that covers the improvement proposition no. 1.

The code change spans 2 projects:
1. Symfony: https://github.com/d-ph/symfony/commit/042cd08fb7743c50fadfb00936b529b3067bbee3
2. DoctrineBundle: https://github.com/d-ph/DoctrineBundle/commit/ae224f31f0dabe1ad77f6996364443ef946fcfb3

Comments:
1. I tested it in my small project and it works better than I expected. Obviously this is subjective, but perhaps worthwhile to mention.
2. The biggest conceptual change is that the new implementation forbids starting a Messenger Worker for queues that are (1) a mixture of postgres and non-postgres, and (2) if the postgres queues use different pg_notify options that are relevant to the implementation (e.g. the pg_notify timeout, or the db table name used by the queue). I'd say that the requirement is not a problem, since it's rare for devs to use a mixture of backend queue solutions, and even if they do use them, then it's still just a matter of starting the Worker with different set of queues.
3. The {{PostgreSqlWaitForPgNotifyOnIdleListener::registerPostgreSqlConnectionCandidate()}} is an annoying necessity stemming from the fact that there is no way to retrieve the {{Symfony\Component\Messenger\Bridge\Doctrine\Transport\Connection}} out of {{DoctrineReceiver}} and {{DoctrineTransport}}. If a getter is added to both of this classes, then the {{::registerPostgreSqlConnectionCandidate()}} may be removed, which I would recommend doing if that's ok. It would also remove the necessity of injecting {{PostgreSqlWaitForPgNotifyOnIdleListener}} to {{DoctrineTransportFactory}}.
4. {{Symfony\Component\Messenger\Bridge\Doctrine\Transport\Connection::$queueEmptiedAt}} is removed. This is a nice side effect of the proposed code change, since the variable was related to an obscure postgres implementation detail.
5. {{PostgreSqlConnection}} now doesn't override the {{get()}} function, which perhaps is a plus.
6. {{PostgreSqlWaitForPgNotifyOnIdleListener::choosePgNotifyTimeout()}} is awaiting the "improvement proposition no. 2" implementation, which I'd say is the "killer feature" of this github ticket.
7. The code change doesn't pass tests and coding standards yet. I'll get it working once there's a confirmation that the code change direction is sound.

Thank you.

**Comment by SherinBloemendaal** [comment]
We would love to see this fix implemented! Thank you! We are facing the exact same issue when consuming multiple queues.

**Comment by carsonbot** [comment]
Thank you for this issue.
There has not been a lot of activity here for a while. Has this been resolved?

**Comment by carsonbot** [comment]
Just a quick reminder to make a comment on this. If I don't hear anything I'll close this.

**Comment by carsonbot** [comment]
Hey,

I didn't hear anything so I'm going to close it. Feel free to comment if this is still relevant, I can always reopen!

**Comment by SherinBloemendaal** [comment]
Still relevant

**Comment by christian-kolb** [comment]
@fabpot @d-ph Has proposed a PR https://github.com/symfony/symfony/pull/47666 here but is still waiting for a confirmation whether and which of the proposed approaches will be approved. Can you confirm for him which one would make it through?

**Comment by SherinBloemendaal** [comment]
Can we re-open this? Would love to see the PR merged😍

**Comment by nicolas-grekas** [comment]
Fixed

**Comment by d-ph** [comment]
For a minor record: the "problem number 1" got fixed, but number 2 and 3 are still happening. At the time of writing this, this was further discussed in here: https://github.com/symfony/symfony/pull/47666#issuecomment-3905383759.

---

Edit: problems 2 and 3 are being addressed by a subsequent PR here: https://github.com/symfony/symfony/pull/63402
